Move visualization.limb into visualization.drawing

mwhv2 opened the pull request:
This is a followup to #6251. This PR moves `visualization.limb` into `visualization.draw` (and its corresponding test) and adds a new example to the Example Gallery that demonstrates the usage of the `draw.equator` and `draw.prime_meridian` functions. Combining the two modules also required the updating of the draw_limb wrapper in mapbase and a small edit to the “Plotting the solar limb” example. 

I have also added two notes to the What’s New 4.1 log that describe the new draw functions and detail the removal of the `visualization.limb` module. On a related note, I am not entirely certain what to label the changelog since there a few things happening here.

nabobalis commented on sunpy/visualization/limb.py:
```diff
@@ -1,99 +0,0 @@
-from sunpy.visualization import wcsaxes_compat
-from sunpy.visualization.draw import _plot_vertices
-
-__all__ = ['draw_limb']
```
We can not delete this wholesale, we will need to go through a deprecation cycle. 
An example is https://docs.sunpy.org/en/latest/dev_guide/contents/public_api.html?highlight=deprecation#deprecation-policy-and-breaking-changes 

You will want to restore the file, either keep the existing function or import the old one if it has the same API and wrap it here.

mwhv2 commented:
I agree that `draw_equator` and `draw_prime_meridian` are better choices for function names. At the moment, I have renamed the functions and changed the module to `visualization.drawing`. 

Before I make these commits, I will wait to hear if the overall consensus is to use the `visualization.drawing` sub-module or move the functions to `sunpy.visualization`.

ayshih commented:
I disagree with the above sentiments*, but I don't care enough to object.  If the verbs are kept, the functions should be grouped in the `draw`/`drawing` module and imported up a level to `sunpy.visualization`.

\* When all of your functions in a module are named `draw_...()` and the name of the module is clear, it is defensible to remove the pointless repetition of verbs.  It's why I removed `get_...` from the fronts of the names of the functions that now live in `sunpy.coordinates.sun`.  There are lots of other examples of groups of functions that all do stuff but skip having verbs in their names, and `draw.limb()` is no less readable or understandable than `draw_limb()`.

nabobalis commented:
> * When all of your functions in a module are named `draw_...()` and the name of the module is clear, it is defensible to remove the pointless repetition of verbs. It's why I removed `get_...` from the fronts of the names of the functions that now live in `sunpy.coordinates.sun`. There are lots of other examples of groups of functions that all do stuff but skip having verbs in their names, and `draw.limb()` is no less readable or understandable than `draw_limb()`.

This is fair, I am not undecided again on what we do. I was thinking that draw_limb and then we hide the `draw` via the imports in the init file was the way to go but looking at that example, the difference is basically the same.

ayshih commented:
The benefit of using `draw.limb()` instead of `draw_limb()` is that it keeps the namespace cleaner without any redundant characters.  When we had only `draw_limb()`, of course we wouldn't separate it out to a module, but if we had ~10 drawing functions, I'd lobby for the `draw.limb()` approach.  At three functions, I consider it a toss-up.

nabobalis commented:
I got the impression we will plan to expand the draw namespace in future.

mwhv2 commented:
> The benefit of using `draw.limb()` instead of `draw_limb()` is that it keeps the namespace cleaner without any redundant characters.

> I got the impression we will plan to expand the draw namespace in future.

Yes, these are both reasons why I originally named the module draw.

nabobalis commented on changelog/6332.feature.deprecation.rst:
```diff
@@ -0,0 +1,4 @@
+Moved :func:`~sunpy.visualization.limb.draw_limb` into `~sunpy.visualization.drawing`
+as :func"`~sunpy.visualization.drawing.limb` and added the plot_equator_prime_meridian.py
+example to the Example Gallery.
+Deprecated the `~sunpy.visualization.limb` module.
```
This will need a separate changelog per item.

nabobalis replied:
You should be able to link the example like you did in the what's new section.

mwhv2 replied:
Okay, would the moving/renaming of `draw_limb` be considered breaking, or just part of the deprecation?

nabobalis commented on docs/whatsnew/4.1.rst:
```diff
@@ -33,6 +33,18 @@ between the consecutive coordinates, in pixel space, and then indexing the data
+Deprecation of `~sunpy.visualization.limb`
+==========================================
+The `~sunpy.visualization.limb` module is deprecated and will be removed in
+version 4.2. The `~sunpy.visualization.limb.draw_limb` function has been moved into
```
```suggestion
version 5.1. The `~sunpy.visualization.limb.draw_limb` function has been moved into 
```

nabobalis commented on examples/plotting/plot_equator_prime_meridian.py:
```diff
@@ -0,0 +1,52 @@
+aia_map.plot()
+drawing.equator(ax, color='blue')
+drawing.prime_meridian(ax, color='red')
+plt.show()
```
I forgot now if we have more than one show in an example, I will need to check.

nabobalis commented on sunpy/visualization/limb.py:
```diff
@@ -1,18 +1,14 @@
 __all__ = ['draw_limb']
 
 
+@deprecated(since="4.1", message="We will be moving this",
+            alternative="sunpy.visualization.drawing")
```
```suggestion
@deprecated(since="4.1", message="This will be moved.",
            alternative="sunpy.visualization.drawing.draw_limb")
```

nabobalis commented on changelog/6332.feature.rst:
```diff
@@ -0,0 +1 @@
+Added the plot_equator_prime_meridian.py example to the Example Gallery.
```
Ideally we want to link to the gallery example if you can look at: https://sphinx-gallery.github.io/stable/advanced.html#cross-referencing

nabobalis commented on changelog/6332.trivial.rst:
```diff
@@ -0,0 +1,2 @@
+Updated mapbase.py and example limb_plotting.py
+to use the :func:`~sunpy.visualization.drawing.limb` function.
```
```suggestion
All internal code now uses :func:`~sunpy.visualization.drawing.limb`.
```

nabobalis commented on docs/whatsnew/4.1.rst:
```diff
@@ -33,6 +33,18 @@ between the consecutive coordinates, in pixel space, and then indexing the data
+============================================
+You can now use the `~sunpy.visualization.drawing` module to draw the solar equator
+and prime meridian (zero Carrington longitude) on Astropy WCS axes. See
+:ref:`plot_equator_prime_meridian.py` for an example.
```
Same as the changelog, if we can ref the gallery example directly.

mwhv2 replied:
In the [What's New 3.1](https://docs.sunpy.org/en/stable/whatsnew/3.1.html) page (for example), there are several examples that are referenced using a fancy-looking "minigallery" command. Is that method of referencing preferred in the What's New documentation?

mwhv2 commented:
I am not sure what exactly is failing in the [build_docs (ubuntu-latest)](https://github.com/sunpy/sunpy/runs/7438301054?check_suite_focus=true#step:9:627) check. The `sunpy.visualization.limb` module and `draw_limb` function it references are still here. Additionally, there are no references to `draw` anymore (I am certain these were changed to `drawing`).

nabobalis commented on docs/whatsnew/4.1.rst:
```diff
@@ -33,6 +33,19 @@ between the consecutive coordinates, in pixel space, and then indexing the data
+
+.. minigallery:: sunpy.visualization.drawing.equator
+
+Deprecation of `sunpy.visualization.limb`
```
Might be better to just not link these and use double ticks instead.

mwhv2 commented:
Could these [warnings](https://github.com/sunpy/sunpy/runs/7453377710?check_suite_focus=true#step:9:649) be coming from code links in old changelogs?

nabobalis commented:
That is very possible, you might need to do a full search for any of those references. The CHANGELOG.rst is probably the first place to start.

dstansby commented on changelog/6251.feature.rst:
```diff
@@ -1,3 +1,3 @@
-:func:`~sunpy.visualization.draw.equator` and
-:func:`~sunpy.visualization.draw.prime_meridian` functions.
+Created the `sunpy.visualization.drawing` module which includes
+:func:`~sunpy.visualization.drawing.equator` and
```
```suggestion
new :func:`~sunpy.visualization.drawing.equator` and
```

dstansby commented on changelog/6332.trivial.rst:
```diff
@@ -0,0 +1 @@
+All internal code now uses :func:`~sunpy.visualization.drawing.limb`.
```
```suggestion
All internal code for limb drawing now uses :func:`~sunpy.visualization.drawing.limb`.
```

dstansby commented on examples/plotting/plot_equator_prime_meridian.py:
```diff
@@ -0,0 +1,52 @@
+=============================================
+
+This example shows how you can draw the solar equator and prime meridian
+(zero Carrington longitude) on Maps as seen by the axes observer.
```
```suggestion
(zero Carrington longitude) on maps.
```

dstansby commented on sunpy/visualization/__init__.py:
```diff
@@ -1,2 +1,3 @@
+from sunpy.visualization.drawing import *
```
I think we don't want to import everything in `drawing` into the `visualization` namesapce?
```suggestion
```

dstansby reviewed (approved):
I'll commit my very minor suggestions, and then merge. Thanks a lot for this PR, looking forward to the new functionality in sunpy 4.1 👍
